docs/frontend-system/architecture/03-extensions.md

```diff
@@ -85,68 +85,292 @@ Note that while the `createExtension` is public API and used in many places, it
+  // ...
+  output: {
+    // ❌ Bad example - outputting values of same type
+    element1: reactElementExtensionDataRef,
+    element2: reactElementExtensionDataRef,
+  },
+  factory() {
+    return {
+      element1: <div>Hello World</div>,
+      element2: <div>Hello World</div>,
+    };
+  },
+});
+```
+
+### Core Extension Data
+
+We provide default `coreExtensionData`, which provides commonly used `ExtensionDataRef` - e.g. for `React.JSX.Element`, `RouteRef` or `AppTheme`. They can be used when creating your own extension. For example, the React Element extension data that we defined above is already provided as `coreExtensionData.reactElement`. For a full list and explanations of all types of core extension data, see the [core extension data reference](#TODO).
+
+### Optional Extension Data
+
+By default all extension data is required, meaning that the extension factory must provide a value for each output. However, it is possible to make extension data optional by calling the `.optional()` method. This makes it optional for the factory function to return a value as part of its output. When calling the `.optional()` method you create a new copy of the extension data reference, it does not mutate the existing reference.
+
+```tsx
+const extension = createExtension({
+  // ...
+  output: {
+    element: coreExtensionData.reactElement.optional(),
+  },
+  factory() {
+    return {
+      element: <img src="./assets/logo.png" />,
+    };
+  },
+});
+```
 
 ## Extension Inputs
 
-<!--
+The Extension Data can be passed up to other extensions through their extension inputs. Similar to the outputs seen before, let's create an example an extension with a extension input:
 
-Introduce the concept of extension inputs and how they use extension data to reference types just like outputs
+```tsx
+const navigationExtension = createExtension({
+  // ...
+  inputs: {
+    // [1]: Input
+    logo: createExtensionInput(
+      {
+        element: coreExtensionData.reactElement,
+      },
+      { singleton: true, optional: true },
+    ),
+  },
+  factory({ inputs }) {
+    return {
+      element: (
+        <nav>{inputs.logo.output?.element ?? <span>Backstage</span>}</nav>
+      ),
+    };
+  },
+  // ...
+});
+```
 
-Show how to use the `createExtensionInput` API to create an input for an extension.
+The input (see [1] above) is an object with the properties `$$type`, `extensionData` and `config`, where `createExtensionInput` helps creating this object by taking the `extensionData` as the first parameter, and the `config` as the second parameter making sure the `$$type` is set correctly. The extension data ensures the input is of the type & format we expect. With `config` the format & type can be defined even more precisely. The extension input can be converted from an array of inputs to just a single input by setting `singleton` to `true`. If the extension input is optional, `optional: true` converts the type to be optional as well.
```

Comment:
```suggestion
The input (see [1] above) is an object that we create using `createExtensionInput`. The first argument is the set of extension data that we accept via this input, and works just like the `output` option. The second argument is optional, and it allows us to put constraints on the extensions that are attached to our input. If the `singleton: true` option is set, only a single extension can attached at a time, and unless the `optional: true` option is set it will also be required that there is exactly on attached extension.
```

The `$$type` is more of an internal thing that I don't think we should mention in docs tbh, unless we specifically describe our system for opaque types somewhere. I updated the rest a bit to to try to highlight the difference between input shape and shape of attached extensions